admin/release.py

```diff
@@ -303,3 +319,178 @@ def publish_docs_main(args, base_path, top_level):
+    """
+    optParameters = [
+        ["target", None, b'clusterhq-yum-repository',
+         "The bucket to upload packages to."],
+        ["build-server", None,
+         b'http://build.clusterhq.com',
+         "The URL of the build-server."],
+    ]
+
+    def parseArgs(self, version):
+        self['version'] = version
+
+
+FLOCKER_PACKAGES = [
+    b'clusterhq-python-flocker',
+    b'clusterhq-flocker-cli',
+    b'clusterhq-flocker-node',
+]
+
+
+def update_repo(rpm_directory, target_bucket, target_key, source_repo,
+                packages):
+    """
+    Update ``target_bucket`` yum repository with ``packages`` from
+    ``source_repo`` repository.
+
+    :param FilePath rpm_directory: Temporary directory to download
+        repository to.
+    :param bytes target_bucket: S3 bucket to upload repository to.
+    :param bytes target_key: Path within S3 bucket to upload repository to.
+    :param bytes source_repo: Repository on the build server to get packages
+        from.
+    :param list packages: List of bytes, each specifying the name of a package
+        to upload to the repository.
+    """
+    # Import these here so that this file can be imported when yum is not
+    # available.
+    from yum import YumBase
+    from pakrat import repo
+
+    rpm_directory.createDirectory()
+    s3 = boto.connect_s3()
+    try:
+        # Does not work if there is a '.' in the name
+        bucket = s3.get_bucket(bucket_name=target_bucket)
+    except boto.exception.S3ResponseError:
+        bucket = s3.create_bucket(bucket_name=target_bucket)
+
+    # Download existing repository
+    for item in bucket.list(prefix=target_key):
+        new_item_path = os.path.join(rpm_directory.path, str(item.key))
+        if not os.path.exists(new_item_path):
+            parent = FilePath(new_item_path).parent()
+            if not parent.exists():
+                parent.makedirs()
+            item.get_contents_to_filename(new_item_path)
+
+    # Download requested packages from source repository
+    base = YumBase()
+    # This is necessary to avoid permission issues with the default,
+    # /var/cache/yum/x86_64/20
+    base.setCacheDir()
+
+    base.repos.disableRepo('*')
+    flocker_repo = base.add_enable_repo(repoid='flocker',
+                                        baseurls=[source_repo])
+
+    base.cleanMetadata()
+
+    # XXX This could be more efficient by only downloading the changed files
+    # https://clusterhq.atlassian.net/browse/FLOC-1506
+    yum_packages = base.pkgSack.returnPackages(repoid=flocker_repo.name,
+                                               patterns=packages)
+    flocker_repo.pkgdir = os.path.join(rpm_directory.path, target_key)
+    base.downloadPkgs(yum_packages)
+
+    # Update repository metadata
+    flocker_repo.pkgdir = os.path.join(flocker_repo.pkgdir, 'repodata')
+    repo.create_metadata(repo=flocker_repo)
+
+    # Upload updated repository
+    for root, dirs, files in os.walk(rpm_directory.path):
```

Comment:
I don't think it is significantly more complex or time consuming to do the more efficient thing, so I don't think it is worth not doing now. That is, even if it takes a bit of time, we should do it now.